test/suite/test_checkpoint27.py

```diff
@@ -0,0 +1,147 @@
+        if ckpt is None:
+            ckpt = 'WiredTigerCheckpoint'
+        cursor = self.session.open_cursor(ds.uri, None, 'checkpoint=' + ckpt)
+        seen = {}
+        for k, v in cursor:
+            if v in seen:
+                seen[v] += 1
+            else:
+                seen[v] = 1
+        self.assertTrue(seen in expected)
+        cursor.close()
+
+    def test_checkpoint_evict_page(self):
+        self.key_format='r'
+        self.value_format='S'
+
+        uri = 'table:checkpoint27'
+        nrows = 10000
+        morerows = 10000
+
+        # Create a table.
+        ds = SimpleDataSet(
+            self, uri, 0, key_format=self.key_format, value_format=self.value_format)
+        ds.populate()
+
+        value_a = "aaaaa" * 100
+        value_b = "bbbbb" * 100
+
+        # Write some data.
+        self.large_updates(uri, ds, nrows, value_a)
+        # Write this data out now so we aren't waiting for it while trying to
+        # race with the later data.
+        self.session.checkpoint()
+
+        # Write some more data, and hold the transaction open.
+        session2 = self.conn.open_session()
+        cursor2 = session2.open_cursor(uri)
+        session2.begin_transaction()
+        for i in range(nrows + 1, nrows + morerows + 1):
+            cursor2[ds.key(i)] = value_b
+
+        # Checkpoint in the background.
+        done = threading.Event()
+        ckpt = checkpoint_thread(self.conn, done)
+        try:
+            ckpt.start()
+
+            # Wait for checkpoint to start before committing.
+            ckpt_started = 0
+            while not ckpt_started:
+                stat_cursor = self.session.open_cursor('statistics:', None, None)
+                ckpt_started = stat_cursor[stat.conn.txn_checkpoint_running][2]
+                stat_cursor.close()
+                time.sleep(1)
+
+            session2.commit_transaction()
+        finally:
+            done.set()
+            ckpt.join()
+
+        # There are two states we should be able to produce: one with the original
+        # data and one with the additional data.
+        #
+        # It is ok to see either in the checkpoint (since the checkpoint could
+        # reasonably include or not include the second txn) but not ok to see
+        # an intermediate state.
+        expected_a = { value_a: nrows }
+        expected_b = { value_a: nrows, value_b: morerows }
+        expected = [expected_a, expected_b]
+
+        # Create a lot of tables to generate a large metadata page.
+        for i in range(0, 2000):
+            temp_uri = 'table:test_checkpoint27_' + str(i)
+            self.session.create(temp_uri, 'key_format={},value_format={}'.format(self.key_format, self.value_format))
+            self.large_updates(uri, ds, 1, value_a)
+            # if i == 1:
+            #     self.session.checkpoint()
+
+        # Dirty the metadata page on the last table.
+        # TODO this doesn't work.
+        # self.session.checkpoint()
```

Comment:
The operation other than a checkpoint that will dirty the metadata are creating new tables (WT_SESSION::create), or changing existing tables (WT_SESSION::alter). You could try harder to interleave the checkpoint cursor operations reads with changes to the metadata. i.e: run the `check` operation more regularly, and while (or occasionally between) creating new tables. You could also create larger metadata by passing `app_metadata` to the `create` call. If you include a long string there, it will pad out the size of the metadata needed for each table.